### PR TITLE
Allowing to exclude menu items from displaying in frontend menus (in menu item)

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
#### Description

This is PR takes the contributions from https://github.com/joomla/joomla-cms/pull/8814 and adds the ability to exclude menu items from displaying in frontend menus.

For more info see https://github.com/joomla/joomla-cms/pull/8814
#### How to test
1. Fresh Joomla install with multilanguage.
2. Apply this patch
3. Go to a menu item that is displayed in the frontend main menu, separator "Link Type" and set the option "Display in Menu Modules" to "No" to exclude the menu item from displaying in the frontend menu.
4. Check in the frontend it those items are not displayed in the main menu.
#### Observations

Suggestions are welcome.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 1]
```diff
@@ -83,6 +83,8 @@ COM_MENUS_ITEM_FIELD_MENU_IMAGE_DESC="Select or upload an optional image to be u
 COM_MENUS_ITEM_FIELD_MENU_IMAGE_LABEL="Link Image"
 COM_MENUS_ITEM_FIELD_MENU_TEXT_DESC="If the optional image is added, adds the menu title next to the image. Default is 'Yes'."
 COM_MENUS_ITEM_FIELD_MENU_TEXT_LABEL="Add Menu Title"
+COM_MENUS_ITEM_FIELD_MENU_SHOW_DESC="Select 'No' if you want to exclude this menu item from displaying in the menu modules that display this menu item parent menu."
+COM_MENUS_ITEM_FIELD_MENU_SHOW_LABEL="Display in Menu Modules"
```
Comment (justinherrin) [inline administrator/language/en-GB/en-GB.com_menus.ini]: I hate to sound picky here, but I think the Joomla dev folks like to see the language strings in proper alpha order.

**Comment by andrepereiradasilva** [comment]
@justinherrin done.
Can you test if everything ok now?

**Comment by justinherrin** [comment]
I have tested this item :white_check_mark: successfully on 3d547b68dbcc4695aa9caa3fac8d2931eddb519b<br /><br />I have successfully tested this PR by using a bunch of different menu types (item alias, menu heading, text separator, external url, articles, tags, etc).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8816">issues.joomla.org/joomla-cms/8816</a>.</sub>

**Comment by Fedik** [comment]
just curious (cannot test now) how it will work if parent menu is "hidden" but the children is visible? ... on first look it will fail

**Comment by andrepereiradasilva** [comment]
@Fedik, you're right i tested right now.

If you hide the parent menu item that menu subitems will show, and they shouldn't.
Have to correct that.

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 237811c13313767b599344b052ecb51a19bb19e2<br /><br />Test OK, simple menu item and parent menu item tested<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8816">issues.joomla.org/joomla-cms/8816</a>.</sub>

**Comment by NLRoosj** [comment]
I have tested this item :white_check_mark: successfully on 237811c13313767b599344b052ecb51a19bb19e2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8816">issues.joomla.org/joomla-cms/8816</a>.</sub>

**Comment by brianteeman** [comment]
RTC - thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8816">issues.joomla.org/joomla-cms/8816</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@anibalsanchez @NLRoosj thanks for testing

**Comment by izharaazmi** [comment]
Currently we suggest people to create separate *menu type* and add such menu items which they don't want to show. This PR sounds better than that.
@andrepereiradasilva Can you please elaborate over what benefits this is supposed to give us? As you are the author, you have better insights than me.

**Comment by andrepereiradasilva** [comment]
In my view the benefits are:
1. You don't have to create an "hidden menu" for hidden menu items. You just put then in the menu you want and they hide the items you want with this option.
2. You can now hide only second, third, etc level menu items levels from the menu keeping the information architecture itself organized (breadcrumb, sitemap, sef uri's, etc) as it should.
3. Usability. IMHO it's simpler for a user to understand this behavior.
ex: A user is in a menu item that he don't want to appear in the site menu ... he just select this new option in the menu item itself and hide it. Simple as that. No *tricks*
4. BC. You can still create an "hidden" menu item as you do before, if you wish so.

**Comment by izharaazmi** [comment]
Great! Point #2 is what cheers me.

Can you also add **Hide** and **Unhide** buttons in the list view just like the publish/unpublish behaviour for quick access. Better if we can add a column too for this.

**Comment by anibalsanchez** [comment]
In fact, I think this option is simpler than separated "hidden menu".  A "hidden menu" is a trick that usually leads to have many assorted menu items, with no cohesion.

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 2]
```diff
@@ -81,6 +81,8 @@ COM_MENUS_ITEM_FIELD_LINK_DESC="Link for this menu."
 COM_MENUS_ITEM_FIELD_LINK_LABEL="Link"
 COM_MENUS_ITEM_FIELD_MENU_IMAGE_DESC="Select or upload an optional image to be used with the menu hyperlink."
 COM_MENUS_ITEM_FIELD_MENU_IMAGE_LABEL="Link Image"
+COM_MENUS_ITEM_FIELD_MENU_SHOW_DESC="Select 'No' if you want to exclude this menu item from displaying in the menu modules that display this menu item parent menu."
+COM_MENUS_ITEM_FIELD_MENU_SHOW_LABEL="Display in Menu Modules"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_menus.ini]: As far as a regular user is concerned there is no need to say Modules
"Display in Menu" is enough

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 3]
```diff
@@ -81,6 +81,8 @@ COM_MENUS_ITEM_FIELD_LINK_DESC="Link for this menu."
 COM_MENUS_ITEM_FIELD_LINK_LABEL="Link"
 COM_MENUS_ITEM_FIELD_MENU_IMAGE_DESC="Select or upload an optional image to be used with the menu hyperlink."
 COM_MENUS_ITEM_FIELD_MENU_IMAGE_LABEL="Link Image"
+COM_MENUS_ITEM_FIELD_MENU_SHOW_DESC="Select 'No' if you want to exclude this menu item from displaying in the menu modules that display this menu item parent menu."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_menus.ini]: Simplified the text and added a note that submenus are also hidden
+COM_MENUS_ITEM_FIELD_MENU_SHOW_DESC="Select 'No' if you want to exclude this menu item from displaying in the menu. Note any submenu items will also be hidden."

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 237811c13313767b599344b052ecb51a19bb19e2<br /><br />Does what it says - a useful addition
Made some suggestions to simplify the text a bit and make it better english
I do still wonder if it would not be better as Show/Hide instead of Yes/No - that would seem more appropriate to me<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8816">issues.joomla.org/joomla-cms/8816</a>.</sub>

**Comment by izharaazmi** [comment]
I think this should be in line with properties like published. Params is
not much appropriate IMHO.
Also see my other comments whether those make sense.

**Comment by andrepereiradasilva** [comment]
@izharaazmi i tought about your comments,
> I think this should be in line with properties like published. Params is not much appropriate IMHO.

I see several problems with that:
1. A menu item can be Published and hidden, or Not Published and hidden so is not a state like Published, Not Published, Trashed.
2. Consistensy: none of the other components have an hidden state.

> Can you also add Hide and Unhide buttons in the list view just like the publish/unpublish behaviour for quick access. Better if we can add a column too in the list view for this.

Like the "featured" button in the articles list view?
I guess it can be done with a new field in the table. But is it really worth it? Will it not create some confusion to new users?
IMHO, that is to much emphasis for a option to hide a menu item in the menu modules.

@brianteeman
> I do still wonder if it would not be better as Show/Hide instead of Yes/No - that would seem more appropriate to me

I put "Yes/No" to use the same as "Add Menu Title"  "Yes/No" in the same tab.
But if you guys think "Show/Hide" (with a colour dropdown?) is better i can change it.

**Comment by izharaazmi** [comment]
Well, sounds great to me.

I still think three things will be very useful for the common/average users (priority top to bottom):

1. Bulk action (using checkboxes) to hide/unhide items.
2. A way to quickly say which menu item is hidden or not, without needing to open/edit and lookup each one.
3. Filter by visibility.

Tell me if you want me to do the PR for (any of) these.

**Comment by izharaazmi** [comment]
Regarding #8814. I see this one has also great usability. Assume some case when we want to render same **menu** and show different set of **menu items** in different menu module positions!
With that we could override/inherit the selections at will. 
Can we have a second opinion @brianteeman?

Great work @andrepereiradasilva.

**Comment by brianteeman** [comment]
Let's stick to the subject of this PR and not discuss other ones

**Comment by andrepereiradasilva** [comment]
> I still think three things will be very useful for the common/average users (priority top to bottom):
> Bulk action (using checkboxes) to hide/unhide items.
> A way to quickly say which menu item is hidden or not, without needing to open/edit and lookup each one.
> Filter by visibility. 

@izharaazmi I will not do more PR regarding this matter (i have other PR to do).

**Comment by wilsonge** [comment]
Merged for 3.5.0 - thanks @andrepereiradasilva :)

**Pull request closed without merging**